File: Back-End/Db/models/reward_model.js
import mongoose, { Schema, Types, model } from "mongoose";

import { queryHelperPlugin } from "../query_helpers.js";

const rewardSchema = new Schema(
  {
    name: {
      type: String,
      required: true,
      trim: true,
    },
    createdBy: {
      type: Types.ObjectId,
      ref: "Parent",
      required: true,
    },
    description: {
      type: String,
    },
    image: {
      url: String,
      id: String,
    },
    points: {
      type: Number,
      min: 1,
      required: true,
    },
    isActive: {
      type: Boolean,
      default: true,
    },

    isDeleted: {
      type: Boolean,
      default: false,
    },

    quantity: {
      type: Number,
      default: 1,
      min: 0,
    },
  },
  { timestamps: true },
);

rewardSchema.plugin(queryHelperPlugin);
const Reward = mongoose.models.Reward || model("Reward", rewardSchema);

export default Reward;

File: Front-End/lib/features/rewards/data/models/reward_model.dart
/// Reward model for `GET reward/rewards` and the add/update responses.
class RewardModel {
  final String id;
  final String name;
  final String description;
  final int points;
  final int quantity;
  final bool isActive;
  final String imageUrl;

  const RewardModel({
    required this.id,
    required this.name,
    required this.description,
    required this.points,
    required this.quantity,
    required this.isActive,
    required this.imageUrl,
  });

  static int _toInt(dynamic v) =>
      v is num ? v.toInt() : (int.tryParse('$v') ?? 0);

  factory RewardModel.fromJson(Map<String, dynamic> json) {
    final image = json['image'];
    return RewardModel(
      id: (json['_id'] ?? '').toString(),
      name: (json['name'] ?? '').toString(),
      description: (json['description'] ?? '').toString(),
      points: _toInt(json['points']),
      quantity: _toInt(json['quantity']),
      isActive: json['isActive'] != false, // default true
      imageUrl: image is Map ? (image['url']?.toString() ?? '') : '',
    );
  }
}
